src/theory/quantifiers/oracle_checker.cpp

```diff
@@ -0,0 +1,125 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Oracle checker
+ */
+
+#include "theory/quantifiers/oracle_checker.h"
+
+#include "expr/node_algorithm.h"
+#include "options/base_options.h"
+#include "smt/env.h"
+#include "theory/rewriter.h"
+
+namespace cvc5::internal {
+namespace theory {
+namespace quantifiers {
+
+bool OracleChecker::checkConsistent(Node app,
+                                    Node val,
+                                    std::vector<Node>& lemmas)
+{
+  Node result = evaluateApp(app);
+  if (result != val)
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    Node lemma = nm->mkNode(kind::EQUAL, result, app);
+    lemmas.push_back(lemma);
+    return false;
+  }
+  return true;
+}
+
+Node OracleChecker::evaluateApp(Node app)
+{
+  Assert(app.getKind() == kind::APPLY_UF);
+  Node f = app.getOperator();
+  Assert(OracleCaller::isOracleFunction(f));
+  // get oracle caller
+  if (d_callers.find(f) == d_callers.end())
+  {
+    d_callers.insert(std::pair<Node, OracleCaller>(f, OracleCaller(f)));
+  }
+  OracleCaller& caller = d_callers.at(f);
+
+  // get oracle result
+  Node ret;
+  int runResult;
+  caller.callOracle(app, ret, runResult);
+  Assert(!ret.isNull());
+  return ret;
+}
+
+Node OracleChecker::evaluate(Node n)
+{
+  // same as convert
+  return convert(n);
+}
+
+Node OracleChecker::postConvert(Node n)
+{
+  Trace("oracle-checker-debug") << "postConvert: " << n << std::endl;
+  // if it is an oracle function applied to constant arguments
+  if (n.getKind() == kind::APPLY_UF
+      && OracleCaller::isOracleFunction(n.getOperator()))
+  {
+    bool allConst = true;
+    for (const Node& nc : n)
+    {
+      if (nc.isConst())
+      {
+        continue;
+      }
+      // special case: assume all closed lambdas are constants
+      if (nc.getKind() == kind::LAMBDA)
```

Comment:
couldn't the lambda be rewritten and then you'd check if it's constant?

Reply:
The point here is that we intentionally don't rely on `isConst` for LAMBDA terms.  The reason is that it is incomplete, there are some lambdas for which `isConst` will never be true.  The SyGuS standard mentions "closed lambda terms" as being the values for function sorts, which this coincides with.

Reply:
I see, ok.